go/internal/initutil/ipfs.go

```diff
@@ -82,105 +111,78 @@ func (m *Manager) getLocalIPFS() (ipfsutil.ExtendedCoreAPI, *ipfs_core.IpfsNode,
 	}
 
 	// Tor
 	var p2pOpts libp2p.Option
 	var ipfsConfigPatch ipfsutil.IpfsConfigPatcher
-	{
-		setListenTor := true
-		// Scan all maddrs to see if an `/onion{,3}/*` were set, if don't auto add the `tor.NopMaddr3Str`.
-		for _, v := range swarmAddrs {
-			maddr, err := ma.NewMultiaddr(v)
-			if err != nil {
-				// This error is not our business and will be dealt later.
-				continue
-			}
-			protos := maddr.Protocols()
-			if len(protos) == 0 {
-				continue
-			}
-			if proto := protos[0].Code; proto == ma.P_ONION3 || proto == ma.P_ONION {
-				setListenTor = false
-				break
+	if m.Node.Protocol.Tor.Mode == "required" {
+		// Patch the IPFS config to make it complient with an anonymous node.
+		ipfsConfigPatch = func(c *ipfs_cfg.Config) error {
+			// Disable IP transports
+			c.Swarm.Transports.Network.QUIC = ipfs_cfg.False
+			c.Swarm.Transports.Network.TCP = ipfs_cfg.False
+			c.Swarm.Transports.Network.Websocket = ipfs_cfg.False
+
+			// Disable MDNS
+			c.Discovery.MDNS.Enabled = false
+
+			// Replace listeners
+			if !hasTorMaddr(swarmAddrs) {
+				c.Addresses.Swarm = []string{tor.NopMaddr3Str}
+			} else {
+				c.Addresses.Swarm = nil
 			}
+			return nil
 		}
-		if m.Node.Preset == PresetAnonymity {
-			// Force tor in this mode
-			m.Node.Protocol.Tor.Enabled = true
-			// Disable proximity communications
-			m.Node.Protocol.LocalDiscovery = false
-			m.Node.Protocol.MultipeerConnectivity = false
-			// Patch the IPFS config to make it complient with an anonymous node.
-			ipfsConfigPatch = func(c *ipfs_cfg.Config) error {
-				// Disable IP transports
-				c.Swarm.Transports.Network.QUIC = ipfs_cfg.False
-				c.Swarm.Transports.Network.TCP = ipfs_cfg.False
-				c.Swarm.Transports.Network.Websocket = ipfs_cfg.False
-
-				// Disable MDNS
-				c.Discovery.MDNS.Enabled = false
-
-				// Replace listens
-				if setListenTor {
-					c.Addresses.Swarm = []string{tor.NopMaddr3Str}
-				} else {
-					c.Addresses.Swarm = nil
-				}
-				return nil
+	}
+	if !m.Node.Protocol.DisableIPFSNetwork {
+		switch m.Node.Protocol.Tor.Mode {
+		case TorOptional, TorRequired:
+			torOpts := torcfg.SetTemporaryDirectory(tempdir.TempDir())
+			if m.Node.Protocol.Tor.BinaryPath == "" {
+				torOpts = torcfg.Merge(torOpts, torcfg.EnableEmbeded)
+			} else {
+				torOpts = torcfg.Merge(torOpts, torcfg.SetBinaryPath(m.Node.Protocol.Tor.BinaryPath))
 			}
-		}
-		if !m.Node.Protocol.DisableIPFSNetwork {
-			if m.Node.Protocol.Tor.Enabled {
-				torOpts := torcfg.SetTemporaryDirectory(tempdir.TempDir())
-				if m.Node.Protocol.Tor.BinaryPath == "" {
-					torOpts = torcfg.Merge(torOpts, torcfg.EnableEmbeded)
-				} else {
-					torOpts = torcfg.Merge(torOpts, torcfg.SetBinaryPath(m.Node.Protocol.Tor.BinaryPath))
-				}
 
-				if setListenTor {
-					swarmAddrs = append(swarmAddrs, tor.NopMaddr3Str)
-				}
+			if !hasTorMaddr(swarmAddrs) {
+				swarmAddrs = append(swarmAddrs, tor.NopMaddr3Str)
+			}
```

Comment:
```suggestion
```
Not required anymore as this is set in the `ipfsConfigPatch` earlier.

Reply:
the ipfsconfigpatch is only used for tor==required, I will try to refactor the code to remove this